monai/data/dataset.py

```diff
@@ -53,6 +61,135 @@ def __getitem__(self, index):
+        for _transform in self.transform.transforms:
+            # execute all the deterministic transforms before the first random transform
+            if isinstance(_transform, Randomizable):
+                break
+            item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _first_random_and_beyond_transform(self, item_transformed):
+        """
+        Process the data from before the first random transform to the final state ready for evaluation.
+        Args:
+            item_transformed: The data to be transformed (already process upto the first random transform)
+        Returns:
+            the transformed element through the random transforms
+        """
+        start_post_randomize_run = False
+        for _transform in self.transform.transforms:
+            if start_post_randomize_run or isinstance(_transform, Randomizable):
+                start_post_randomize_run = True
+                item_transformed = apply_transform(_transform, item_transformed)
+        return item_transformed
+
+    def _pre_first_random_cachecheck(self, item_transformed):
+        """
+            A function to cache the expensive input data transform operations
+            so that huge data sets (larger than computer memory) can be processed
+            on the fly as needed, and intermediate results written to disk for
+            future use.
+        Args:
+            item_transformed: The current data element to be mutated into transformed representation
+
+        Returns:
+            The transformed data_element, either from cache, or explicitly computing it.
+
+        Warning:
+            The current implementation does not encode transform information as part of the
+            hashing mechanism used for generating cache names.  If the transforms applied are
+            changed in any way, the objects in the cache dir will be invalid.  The hash for the
+            cache is ONLY dependant on the input filename paths.
+        """
+        if item_transformed.get("cached", False) is False:
+            hashfile = None
+            if self.cache_dir is not None:
+                cache_dir_path: Path = Path(self.cache_dir)
+                if cache_dir_path.is_dir():
+                    # TODO: Find way to hash transforms content as part of the cache
+                    data_item_md5 = hashlib.md5(json.dumps(item_transformed, sort_keys=True).encode('utf-8')).hexdigest()
+                    hashfile: Path = Path(cache_dir_path) / f"{data_item_md5}.pt"
+
+            if hashfile is not None and hashfile.is_file():
+                item_transformed = torch.load(hashfile)
+            else:
+                item_transformed = self._pre_first_random_transform(item_transformed)
+                if hashfile is not None:
+                    # add sentinal flag to indicate that the transforms have already been computed.
+                    item_transformed["cache"] = True
```

Comment:
Should we change the flag to True when the `save` operation is really done?

Reply:
This needs to be set before writing to cache so that if the item is restored from the cache the value is present and set.